docs/DynamicCasting.md

```diff
@@ -0,0 +1,683 @@
+See the discussion of Array casting below for a specific case where this non-determinism can be important.
+
+The following sections detail the rules that govern the casting operations for particular Swift types.
+Except as described below, casting between dissimilar types always fails -- for example, casting a struct instance to a function type or vice-versa.
+
+Where possible, each section includes machine-verifiable _invariants_ that can be used as the basis for developing a robust test suite for this functionality.
+
+## Identity Cast
+
+Casting an instance of a type to its own type will always succeed and return the original value unchanged.
+
+```
+let a: Int = 7
+a is Int // true
+a as? Int // Succeeds
+a as! Int == a // true
+```
+
+## Classes
+
+Casting among class types follows standard object-oriented programming conventions:
+
+* Class upcasts: If `C` is a subclass of `SuperC` and `c` is an instance of `C`, then `c is SuperC == true` and `(c as? SuperC) != nil`.  These "upcasts" do not change the representation in any way.  However, when `c` is accessed via a variable or expression of type `SuperC`, only methods and instance variables defined on `SuperC` are available.
+
+* Class downcasts: If `C` is a subclass of `SuperC` and `sc` is an instance of `SuperC`, then `sc is C` will be true iff `sc` is actually an instance of `C`.  When this "downcast" does succeed, it does not change the representation.
+
+* Objective-C class casting:  The rules above also apply when one of the classes in question is defined with the `@objc` attribute or inherits from an Objective-C class.
+
+* Class casts to AnyObject: Any class reference can be cast to `AnyObject` and then cast back to the original type.  See "AnyObject" below.
+
+* If a struct or enum type conforms to `_ObjectiveCBridgeable`, then classes of the associated bridging type can be cast to the struct or enum type and vice versa.  See "The `_ObjectiveCBridgeable` Protocol" below.
+
+Invariants:
+* For any two class types `C1` and `C2`: `c is C1 && c is C2` iff `((c as? C1) as? C2) != nil`
+* For any class type `C`: `c is C` iff `(c as! AnyObject) is C`
+* For any class type `C`: if `c is C`, then `(c as! AnyObject) as! C === c`
+
+## Structs and Enums
+
+You cannot cast between distinct struct or enum types.
+More formally:
+
+* If `S` and `T` are struct or enum types and `s is S == true`, then `s is T` iff `S.self == T.self`.
+
+## Tuples
+
+Casting from a tuple type T1 to a tuple type T2 will succeed iff the following hold:
+* T1 and T2 have the same number of elements
+* If an element has a label in both T1 and T2, the labels are identical
+* Each element of T1 can be individually cast to the corresponding type of T2
+
+## Functions
+
+Casting from a function type F1 to a function type F2 will succeed iff the following hold:
+* The two types have the same number of arguments
+* Corresponding arguments have identical types
+* The return types are identical
```

Comment:
Remind me again, are covariant/contravariant optionals supported for arguments/return values?

Reply:
No.